functable.c

```diff
@@ -157,3 +195,51 @@ ZLIB_INTERNAL uint32_t crc32_stub(uint32_t crc, const unsigned char *buf, uint64
 
     return functable.crc32(crc, buf, len);
 }
+
+ZLIB_INTERNAL int32_t compare258_stub(const unsigned char *src0, const unsigned char *src1) {
+#ifdef UNALIGNED_OK
+#  ifdef HAVE_BUILTIN_CTZLL
+    functable.compare258 = &compare258_unaligned_64;
+#  elif defined(HAVE_BUILTIN_CTZ)
+    functable.compare258 = &compare258_unaligned_32;
+#  else
+    functable.compare258 = &compare258_unaligned_16;
+#  endif
+#  if defined(X86_AVX2) && defined(HAVE_BUILTIN_CTZ)
+    if (x86_cpu_has_avx2)
+        functable.compare258 = &compare258_unaligned_avx2;
+#  endif
+#  ifdef X86_SSE42_CMP_STR
```

Comment:
AVX2 version should come after SSE4.2 version as otherwise it will never be used.

Reply:
In my tests SSE4.2 was a bit faster.

Reply:
That only means we have to try to optimize it further...

Reply:
```
static inline int32_t compare258_unaligned_avx2_static(const unsigned char *src0, const unsigned char *src1) {
    const unsigned char *src0start = src0;
    const unsigned char *src0end = src0 + 256;
 
    do {
        __m256i ymm_src0, ymm_src1, ymm_cmp;
        ymm_src0 = _mm256_loadu_si256((__m256i*)src0);
        ymm_src1 = _mm256_loadu_si256((__m256i*)src1);
        ymm_cmp = _mm256_cmpeq_epi8(ymm_src0, ymm_src1); // non-identical bytes = 00, identical bytes = FF
        int mask = _mm256_movemask_epi8(ymm_cmp); 
        if ((unsigned int)mask != 0xFFFFFFFF) {
            int match_byte = __builtin_ctz(~mask); // Invert bits so identical = 0
            return (int32_t)(src0 - src0start + match_byte);
        }

        src0 += 32, src1 += 32;
    } while (src0 < src0end);

    if (*(uint16_t *)src0 == *(uint16_t *)src1)
        src0 += 2, src1 += 2;
    else if (*src0 == *src1)
        src0 += 1, src1 += 1;

    return (int32_t)(src0 - src0start);
}
```

This version doesn't use xor at all... basically it compares the operands directly... this works because we assume all ones is the best case scenario instead of all zeroes.

Reply:
Now AVX2 matches or beats SSE4.2

Reply:
That's good... It just took little extra effort due to AVX2 not having cmpne... Reversing the tests and using GPR-version of NOT was the best solution... AVX512 has cmpne, so it will improve performance even more when more machines have it...

Reply:
It believe it should be: `__builtin_ctz(~mask) / 8`.

Reply:
Nah... the _mm256_movemask_epi8() already strips the extra 7 bits...